### PR TITLE
Start of replacing submitted third party libraries with npm modules

### PR DESCRIPTION
**Pull request by mramato**
NOTE: This is a proof-of-concept PR to discuss changing our overall approach to third party libraries.

Rather than submit libraries to `Source/ThirdParty`, which always end up being modified for our build system and quickly go out of date, this change starts to use libraries via npm instead. Currently Autolinker, earcut, when, tween.js, rbush, kdbush, quickselect, and topojson are ported.

The main hurdle that prevented us from doing this sooner was the fact that Cesium has a long history of not requiring a build step after every code change and has a goal of writing valid JS code, unlike many libraries today that mandate a bundler to turn invalid JS code into valid code.

Rather than mandate a bundler during development, this initial step adds a "buildThirdyParty" function to the current "build" step. This function runs third party libraries (defined in `ThirdParty/ThirdParty.js` through RollUp and creates a single' `Source/ThirdParty/ThirdParty.js` ES module that we can use in our own code. The change to end users will be non-existent, especially since the combined Cesium.js will still re-export any third party modules as part of the private API just like it used to.

This doesn't prevent code duplication for users using some of the same third party libraries as Cesium. This is just an improvement as to how depend on third party libraries internally.

I think Cesium's days of being "bundler free" are probably limited long term, performance is the main hurdle and newer tools like esbuild may make that no longer a problem. But that's outside the scope of this initial goal.

Currently this branch won't work on IE11 because the third party code isn't run through babel to transpile it down. IE11's days are numbered as well, so we'll have to decide what we want to do. It may be easy to just use the pre-transpiled versions of third party libraries with this technique and support IE11 without babel for the time being.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @mramato!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/npm-third-party/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

**Comment by ptrgags** [comment]
@mramato Just a drive-by comment because I'm curious. So far this looks like a step in the right direction, I'm able to run sandcastle examples just fine in Chrome at least.

>The main hurdle that prevented us from doing this sooner was the fact that Cesium has a long history of not requiring a build step after every code change and has a goal of writing valid JS code, unlike many libraries today that mandate a bundler to turn invalid JS code into valid code.

For those of us that don't know the full history of CesiumJS, why was not requiring a build step such a priority? I will say it is nice to not have to run `npm run build` most of the time. But nowadays, most build tools and transpilers have a `--watch` flag, so I don't see why it would be a huge concern anymore except if the watching build is so slow it becomes cumbersome for local development. Are there other reasons for avoiding a build step?

**Comment by mramato** [comment]
> watching build is so slow it becomes cumbersome for local development

This is the primary reason. Iteration goes from near instant to ~10-15 seconds every time you change a line of code. It needs to be so fast that the developer doesn't even notice.

> Are there other reasons for avoiding a build step?

1. Philosophically, we want to write JavaScript, not something that is 95% JavaScript but invalid.  At that point we become entirely dependant on the tools. This is not the best reason anymore, but it is a shift from our original approach so I wanted to mention it.
2. We don't want to require our users use bundlers to use our modules (though this is becoming more commonplace and we might end up being okay with it in the long run)
3. Historically, most of the tooling simply wasn't there that could handle seems necessary complexity (dealing with workers, coverage, debugging, etc..).  But it's just that _historical_.

So long story short, it's just a big paradigm shift for how we develop and someone needs to dot all the i's and cross all of the t's before we get there (similar to what we did for initial ES6 module support).  It will happen, just not overnight.

**Comment by lilleyse** [comment]
Couple of topics that came up during offline discussion:

* @ebogo1 is going to open an issue with the full plan for updating third party libraries and a checklist of what's left
* This PR should target a staging branch since this and future third party updates won't work in IE11.
* Is it a problem that all the npm third party libraries are bundled into a single file `ThirdParty/ThirdParty.js`? Or is it not a problem because an app's build system can take advantage of tree shaking? How does pulling in all of `ThirdParty.js` affect workers?

**Comment by ebogo1** [comment]
@mramato now that this targets a staging branch, can it be marked ready for review and merged?

**Comment by mramato** [comment]
> * Is it a problem that all the npm third party libraries are bundled into a single file `ThirdParty/ThirdParty.js`? Or is it not a problem because an app's build system can take advantage of tree shaking? How does pulling in all of `ThirdParty.js` affect workers?

Excellent question. Rollup treeshaking is supposed to take care of this. If you check out this branch and run `npm run build` you can search the `Workers` directory for "earcut" and you'll see it is included, but if you search for `topojson` you'll see it's nowhere to be found.  This is because topojson isn't actually used by any of our works and is therefore not included.

However, treeshaking appears to be imperfect.  AutoLinker.js s not included in any worker but does get included in the build of this branch.  To play is safe, I'm going to break this up into individual files instead of a single ThirdParty.js fo now (But would like to revisit in the future at some point).

**Comment by mramato** [comment]
Okay, I retooled this to be what I described above.

* For every npm module we depend on directly, there is now an equivalent `ThirdParty/npm/` JS file that imports and re-exports it.
* rollup builds these into a temporary `Source/ThirdPartyNpm` directory (once the refactor is complete, we should rename this back to `ThirdParty`)
* our npm modules will share code if they have their own common dependencies
* Calling code is unchanged (except for temporary file location change), and end user is unchanged as well.

@lilleyse not sure if you'd like to take another look.  @ebogo1 I'm fine with merging this into the staging branch (once CI passes) assuming we are all on the same page regarding this approach.

**Comment by ebogo1** [comment]
> our npm modules will share code if they have their own common dependencies

where does this happen?

**Comment by mramato** [comment]
> where does this happen?

In `buildThirtdParty`, it's an "free" side effect of using rollup.
